Deepomatic/dmake
# Add DMake internal plan graph tests

Test several dmake commands, and compare expected internal graph (via
the `--debug-graph` feature).

DMake generates an internal graph planning the various steps to
execute, according to the command-line and the dmake.yml files.

Example:
service foo with needed_service bar, `dmake test foo` execution plan
graph would have the following edges:
- test foo -> build foo
- test foo -> run bar
- build foo -> build-base foo
- run bar -> test bar
- run bar -> build bar
- test bar -> build bar
- build bar -> build-base bar

This graph can be explored with `dmake --debug-graph` options.

To make execution plan construction easier when adding features to
DMake, this PR adds tests comparing expected with actual execution
plan graph dumps.

The reference graphs can be generated with:
```
make -C test rebuild_graphs
```

Required some cleanup:
- split `argparser` construction out of `dmake` and into `cli.py` to
  be importable
- introduce deepbuild.reset() to reset static classes
  states (just SharedVolumes for now)
- stabilize graph dump order: both ordered dump and ordered
  NeededServiceSerializer env* str/repr.

Also:
- less sys.exit()
- more verbose pytest output to show full diffs on error
- fix/improve test/ project, notably add 2 needed_services on same
  service, but different specialization, to test an edge case on
  NeededServiceSerializer sort.
